### PR TITLE
migrate to golang.zx2c4.com/wireguard/wgctrl

### PR DESCRIPTION
**Pull request by leonnicolas**
This commit introduces the usage of wgctrl.
It avoids the usage of exec calls of the wg command
and parsing the output of `wg show`.

### CODE REVIEW

**Comment by squat** [comment]
could you split the golang vendoring into its own separate commit so we can review the changes to Kilo separately and more clearly?

**Comment by leonnicolas** [comment]
OK, the first e2e test failed because I forgot that we were using wg from the kilo container and the kilo container doesn't have wg-tools anymore.

**Comment by squat** [comment]
> OK, the first e2e test failed because I forgot that we were using wg from the kilo container and the kilo container doesn't have wg-tools anymore.

:+1: aha nice

File: pkg/mesh/backend.go [hunk 1]
```diff
@@ -54,8 +56,9 @@ const (
 
 // Node represents a node in the network.
 type Node struct {
-	Endpoint     *wireguard.Endpoint
-	Key          []byte
+	KiloEndpoint *wireguard.Endpoint
```
Comment (squat) [inline pkg/mesh/backend.go:59]: This new abstraction is pretty confusing to me. Why do nodes and peers have to have both a `KiloEndpoint` and an `Endpoint`? It seems like it will be very easy to forget to do some bookkeeping with these two pointers and break things. Semantically they are also very confusing because they are both "endpoints that Kilo uses", no? Maybe it would be easier to understand everything if we instead had:
```golang
type Node struct {
    DNS string
    Endpoint *net.UDPAddr
}
```

Comment (squat) [reply pkg/mesh/backend.go:59]: Also, is the `Endpoint` field used outside of this package or can it be turned into a private field?

Comment (leonnicolas) [reply pkg/mesh/backend.go:59]: If save DNS in a separate string, where would we save the port? Do you mean an addr field like `<dns name>:<port>`?

File: cmd/kgctl/connect_linux.nogo [hunk 2]
```diff
@@ -0,0 +1,473 @@
+// +build linux
```
Comment (squat) [inline cmd/kgctl/connect_linux.nogo:1]: This is supposed to be in another pull request, right?

File: cmd/kgctl/connect_other.nogo [hunk 3]
```diff
@@ -0,0 +1,20 @@
+// +build !linux
```
Comment (squat) [inline cmd/kgctl/connect_other.nogo:1]: This also, no?

File: cmd/kgctl/showconf.go [hunk 4]
```diff
@@ -151,14 +152,14 @@ func runShowConfNode(_ *cobra.Command, args []string) error {
 		}
 	}
 
-	t, err := mesh.NewTopology(nodes, peers, opts.granularity, hostname, opts.port, []byte{}, subnet, nodes[hostname].PersistentKeepalive, nil)
+	t, err := mesh.NewTopology(nodes, peers, opts.granularity, hostname, int(opts.port), wgtypes.Key{}, subnet, nodes[hostname].PersistentKeepalive, nil)
```
Comment (squat) [inline cmd/kgctl/showconf.go:155]: Do we want to directly redefine the port as an int so we don't have to convert from uint? If we want to limit values we can always add a validation function

File: cmd/kgctl/connect_other.nogo (hunk 3 above)
Comment (leonnicolas) [reply cmd/kgctl/connect_other.nogo:1]: yes, i was looking for that file. I thought i deleted it!!

Comment (squat) [reply cmd/kgctl/connect_other.nogo:1]: Agh that would have been a bummer 🙀

File: pkg/k8s/backend.go [hunk 5]
```diff
@@ -367,53 +370,57 @@ func translatePeer(peer *v1alpha1.Peer) *mesh.Peer {
...
 	}
-	var key []byte
-	if len(peer.Spec.PublicKey) > 0 {
-		key = []byte(peer.Spec.PublicKey)
-	}
-	var psk []byte
-	if len(peer.Spec.PresharedKey) > 0 {
-		psk = []byte(peer.Spec.PresharedKey)
+
+	key, _ := wgtypes.ParseKey(peer.Spec.PublicKey)
```
Comment (squat) [inline pkg/k8s/backend.go:401]: What about this error?

File: cmd/kgctl/showconf.go (hunk 4 above)
Comment (leonnicolas) [reply cmd/kgctl/showconf.go:155]: Maybe it makes more sense, because ports can't be higher than 2**16 and we didn't check this before.

File: pkg/k8s/backend.go (hunk 5 above)
Comment (leonnicolas) [reply pkg/k8s/backend.go:401]: Should we log it? We were ignoring it before (or we didn't know if this was a valid key)

File: cmd/kgctl/connect_linux.nogo (hunk 2 above)
Comment (leonnicolas) [reply cmd/kgctl/connect_linux.nogo:1]: yes

File: pkg/mesh/backend.go [hunk 6]
```diff
@@ -54,8 +56,9 @@ const (
 
 // Node represents a node in the network.
 type Node struct {
-	Endpoint     *wireguard.Endpoint
-	Key          []byte
+	Endpoint     *net.UDPAddr
+	Addr         string // eg. dnsname:port
```
Comment (squat) [inline pkg/mesh/backend.go:60]: Before this change, we only stored three values: IP, DNS, and Port. If we keep this as Addr, then we are duplicating data, which leads to ambiguity IMO about which is the "real" port. Can we instead make this DNS? Also, if it really needs to have both fields, I think it would be better to make it a struct so we don't always have to split the string. WDYT?
```suggestion
	DNS         string
```

Comment (leonnicolas) [reply pkg/mesh/backend.go:60]: I think it it is awkward to use the net.UDPAddr to only store the port, this would mean we would always have to check if the IP is nil, to see if the endpoint is invalid instead of just `updAddr != nil`.

If we split the dns name and port, more code would be neccesary to resolve the endpoint because right now, we can simply use `net.UDPResolve` which wants a string with IP and port. I think we only split the string once, when we check if it is valid. And we would still do this because it comes form the `kilo.squat.ai/force-endpoint` annotation for nodes or the `v1alpha1` API. So right now, I don't see the benefit of using a new struct with a String method.

Or we continue using the old `wireguard.Enpoint` and translate this to net.UDPAddr at some point in the code before we use compare old and new config and set the config. This feels like going backwards to me, but maybe it makes sense.

Comment (squat) [reply pkg/mesh/backend.go:60]: I worried that we will somehow end up with different values for the ports in each field and that there will be some hard-to-track bugs as a result. This smells a bit to me like a leaky abstraction. Also, it's still confusing to me that we call this field Addr, when it is really only supposed to represent an optional DNS name.
Instead of ResolveUPDAddr, we can do a DNS lookup, which that func does under the hood anyways. That's actually exactly what we already so today: https://github.com/squat/kilo/blob/main/pkg/mesh/mesh.go#L656

File: cmd/kgctl/showconf.go (hunk 4 above)
Comment (leonnicolas) [reply cmd/kgctl/showconf.go:155]: ok added a check for the valid port range 1-2**16-1

File: pkg/mesh/mesh.go [hunk 7]
```diff
@@ -537,10 +535,12 @@ func (m *Mesh) applyTopology() {
...
 		// Setting the WireGuard configuration interrupts existing connections
 		// so only set the configuration if it has changed.
-		equal := conf.Equal(oldConf)
+		conf := t.Conf()
+		equal, diff := conf.Equal(wgDevice)
 		if !equal {
-			level.Info(m.logger).Log("msg", "WireGuard configurations are different")
-			if err := wireguard.SetConf(link.Attrs().Name, confPath); err != nil {
+			level.Info(m.logger).Log("msg", "WireGuard configurations are different", "diff", diff)
+			level.Debug(m.logger).Log("changing wg config", "config", conf.WGConfig())
```
Comment (leonnicolas) [inline pkg/mesh/mesh.go:542]:
```suggestion
			level.Debug(m.logger).Log("msg","changing wg config", "config", conf.WGConfig())
```

File: pkg/mesh/mesh.go [hunk 8]
```diff
@@ -700,18 +686,22 @@ func peersAreEqual(a, b *Peer) bool {
 	}
 	// Check the DNS name first since this package
 	// is doing the DNS resolution.
-	if !a.Endpoint.Equal(b.Endpoint, true) {
+	if a.Endpoint.StringOpt(false) != b.Endpoint.StringOpt(false) {
```
Comment (leonnicolas) [inline pkg/mesh/mesh.go:689]:
```suggestion
	if a.Endpoint.StringOpt(true) != b.Endpoint.StringOpt(true) {
```

File: pkg/k8s/backend.go (hunk 5 above)
Comment (squat) [reply pkg/k8s/backend.go:398]: Yeah i think that's a good idea

File: pkg/mesh/topology.go [hunk 9]
```diff
@@ -267,45 +282,52 @@ CheckIPs:
...
 		return endpoint
 	}
-	e, ok := t.discoveredEndpoints[string(key)]
+	e, ok := t.discoveredEndpoints[key.String()]
 	if ok {
-		return e
+		return wireguard.NewEndpointFromUDPAddr(e)
 	}
-	return endpoint
+	return nil
```
Comment (leonnicolas) [inline pkg/mesh/topology.go:294]:
```suggestion
	return endpoint
```

File: pkg/mesh/topology.go [hunk 10]
```diff
@@ -314,51 +336,58 @@ func (t *Topology) Conf() *wireguard.Conf {
 
 // AsPeer generates the WireGuard peer configuration for the local location of the given Topology.
 // This configuration can be used to configure this location as a peer of another WireGuard interface.
-func (t *Topology) AsPeer() *wireguard.Peer {
+func (t *Topology) AsPeer() wireguard.Peer {
```
Comment (leonnicolas) [inline pkg/mesh/topology.go:339]: Should this return a pointer again?

File: pkg/mesh/topology.go [hunk 11]
```diff
@@ -314,51 +336,58 @@ func (t *Topology) Conf() *wireguard.Conf {
...
 	}
-	return nil
+	return wireguard.Peer{}
 }
 
 // PeerConf generates a WireGuard configuration file for a given peer in a Topology.
-func (t *Topology) PeerConf(name string) *wireguard.Conf {
-	var pka int
-	var psk []byte
+func (t *Topology) PeerConf(name string) wireguard.Conf {
```
Comment (leonnicolas) [inline pkg/mesh/topology.go:357]: Should this return a pointer again?

File: pkg/mesh/topology.go [hunk 12]
```diff
@@ -379,13 +408,13 @@ func findLeader(nodes []*Node) int {
 	var leaders, public []int
 	for i := range nodes {
 		if nodes[i].Leader {
-			if isPublic(nodes[i].Endpoint.IP) {
+			if isPublic(nodes[i].Endpoint.IP()) {
```
Comment (leonnicolas) [inline pkg/mesh/topology.go:411]:
```suggestion
			if !net.IsPrivate(nodes[i].Endpoint.IP()) {
```

Comment (squat) [reply pkg/mesh/topology.go:411]: Sgtm, I think this is relatively new to the stdlib so nice to switch to it

File: pkg/mesh/mesh.go (hunk 8 above)
Comment (squat) [reply pkg/mesh/mesh.go:689]: Hmm I'm not familiar with this lib

File: pkg/mesh/topology.go [hunk 13]
```diff
@@ -379,13 +408,13 @@ func findLeader(nodes []*Node) int {
...
 				return i
 			}
 			leaders = append(leaders, i)
 
 		}
-		if isPublic(nodes[i].Endpoint.IP) {
+		if nodes[i].Endpoint != nil && isPublic(nodes[i].Endpoint.IP()) {
```
Comment (leonnicolas) [inline pkg/mesh/topology.go:417]:
```suggestion
		if nodes[i].Endpoint.IP() != nil && !nodes[i].Endpoint.IP().IsPrivate() {
```

File: pkg/mesh/topology.go (hunk 10 above)
Comment (squat) [reply pkg/mesh/topology.go:339]: Up to you! I think so, that way we can avoid copies when we pass it around/use the methods on it

File: pkg/mesh/topology.go (hunk 11 above)
Comment (squat) [reply pkg/mesh/topology.go:357]: I think so. What was the reason for changing it?

File: pkg/mesh/mesh.go (hunk 8 above)
Comment (leonnicolas) [reply pkg/mesh/mesh.go:689]: Maybe it is not a good idea to compare strings here. Anyways the comment above suggests to check DNS names first, so this is the boolean flag about. Maybe `wireguard.Endpoint` should also implement `Equal(dnsFirst)`?

Comment (squat) [reply pkg/mesh/mesh.go:689]: It used to implement this, right? So you mean un-remove the functionality? I think that could be helpful

File: pkg/mesh/topology.go (hunk 12 above)
Comment (leonnicolas) [reply pkg/mesh/topology.go:411]: we would have to bump go version first :( 
Didn't know it was new.

Comment (squat) [reply pkg/mesh/topology.go:411]: Cool cool, we can always do it in a follow-up

File: pkg/mesh/topology.go (hunk 11 above)
Comment (leonnicolas) [reply pkg/mesh/topology.go:357]: I don't know. I will add a check to `(*Conf).Bytes()` to return `nil, nil` in case `*Conf == nil`.

File: pkg/mesh/topology.go (hunk 10 above)
Comment (leonnicolas) [reply pkg/mesh/topology.go:339]: I changed it back, but then callers of this func have to check if it returned nil.

Comment (squat) [reply pkg/mesh/topology.go:339]: ack

File: pkg/mesh/mesh.go (hunk 8 above)
Comment (leonnicolas) [reply pkg/mesh/mesh.go:689]: We are using it in nodesAreEqual and peersAreEqual.
I think comparing the strings is save. So the Equal func could just call the String method.
Is this a good trade off for adding another function to wireguard.Endpoint?

Comment (squat) [reply pkg/mesh/mesh.go]: yeah, I think it's worth the complexity to make it easier for the callers so that they don't need to remember to check the DNS equality before the IP equality

**Comment by leonnicolas** [comment]
Should also resolve the issue from #140

**Comment by squat** [comment]
> Should also resolve the issue from #140

That's the same thing you saw on a cluster, right?

**Review by squat — APPROVED** [review]

**Comment by squat** [comment]
🌮 🌮
